src/amoc_code_server.erl

```diff
@@ -0,0 +1,279 @@
+%%==============================================================================
+%% Copyright 2023 Erlang Solutions Ltd.
+%% Licensed under the Apache License, Version 2.0 (see LICENSE file)
+%%==============================================================================
+-module(amoc_code_server).
+%% API
+-export([start_link/0,
+         add_module/1,
+         distribute_modules/1,
+         does_scenario_exist/1,
+         list_scenario_modules/0,
+         list_configurable_modules/0]).
+
+-ifdef(TEST).
+
+-export([upload_module/2,
+         uploaded_module_to_map/1,
+         map_to_uploaded_module/1]).
+
+-define(RECORD2MAP(RecordName),
+        (fun(#RecordName{} = Record) ->
+             RecordFields = lists:zip(lists:seq(2, record_info(size, RecordName)),
+                                      record_info(fields, RecordName)),
+             BasicMap = #{'$RECORD_NAME' => RecordName},
+             lists:foldl(fun({FieldPos, FieldName}, Map) ->
+                             Map#{FieldName => erlang:element(FieldPos, Record)}
+                         end,
+                         BasicMap, RecordFields)
+         end)).
+
+-define(MAP2RECORD(RecordName),
+        (fun(#{'$RECORD_NAME' := RecordName} = Map) ->
+             BasicRecord = #RecordName{},
+             RecordFields =
+                 lists:zip3(lists:seq(2, tuple_size(BasicRecord)),
+                            record_info(fields, RecordName),
+                            tl(tuple_to_list(BasicRecord))),
+             lists:foldl(fun({FieldPos, FieldName, DefaultValue}, Record) ->
+                             Value = maps:get(FieldName, Map, DefaultValue),
+                             setelement(FieldPos, Record, Value)
+                         end,
+                         BasicRecord, RecordFields)
+         end)).
+
+-endif.
+
+-behaviour(gen_server).
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2]).
+
+%% renaming the fields of the uploaded_module record will
+%% break most test cases in amoc_code_server_SUITE because
+%% the module info map used in the suite must have key names
+%% matched to the field names of this record.
+-record(uploaded_module, {module :: module(),
+                          beam_file :: file:filename(),
```

Comment:
Maybe this should be named `beam_filepath` or `beam_filename`? `beam_file` almost sounds like the binary contents of the file 🤔